### PR TITLE
Working group: make apply opening stake mandatory and store relation between application and opening

### LINKED ISSUE #2196: Pallets where there is no deposit charged for filling up the storage
## Problem

It was found during the audit that any forum user can use the forum's extrinsic `create_thread` to fill up the blockchain storage and exhaust the number of threads of a given category for a small fee since this requires no deposit.

This problem exists in two forms across the runtime: Either we allow with no deposit filling up the storage of a given module with no bound or with bounds.

In the bounded case, this would allow a denial of service attack where other users won't be able to use the storage. 

In the unbounded case, this'd allow to grow the storage maliciously. 

There are different permission levels that'd allow a user to abuse these extrinsics, either a **lead** or a **worker**/**member**. 

When a **lead** can fill up a **bounded** storage is not really a problem because there is no other user that it can be denied of service but I leave those cases in here just to keep them in mind.

## Pallets

- `forum`
  - `create_thread`: **Any forum user** can create a thread, the thread storage for a given category is **bounded**.
  - `add_post`: **Any forum user** can create a post, the post storage for a given thread is **bounded**.
  - `create_category`: **The lead** can create categories with no cost, the category storage is **bounded**. (Same goes for sub-categories)

- `working_group`
  - `add_opening`: **A lead** can use this extrinsic to create openings **unbounded**.
  - `apply_on_opening`: **A working group member** can apply multiple times, filling blockchain storage(when no stakes) **unbounded**.
  - `add_opening` + `apply_on_opening` + `fill_opening`: **A lead** can add an opening with no stakes, apply on it multiple times and fill it, and occupy completely the worker storage.

- `membership`
  - `add_staking_account_candidate`: **A user** could use multiple accounts to fill up the blockchain storage, **unbounded**. 

- `proposals_discussion`
  - `add_post`: **A member** can add posts to any thread, **unbounded**.

- `storage/data_directory`
  - `add_content`: **A member** can use it to fill up the storage **unbounded**.

- `content_directory`
  - `add_curator_group`: **A lead** can add groups to fill up the storage **unbounded**.
  - `add_curator_to_group`: **A lead** can add curators to a group to fill it up **bounded**.

- `service_discovery`:
  - `set_ipns_id`: **Any worker** can use it to fill up storage **unbounded**.

- `data_object_type_registry`
   - `register_data_object_type`: **A lead** can use it to fill up storage **unbounded**.

### LINKED ISSUE #2216: There is no stored relation between an opening and an application in the working group
## The problem
When a member applies for a given opening the [application doesn't store any information about the opening originally applied to](https://github.com/Joystream/joystream/blob/olympia/runtime-modules/working-group/src/lib.rs#L454-L460).

This means, that later on, when [filling the opening](https://github.com/Joystream/joystream/blob/olympia/runtime-modules/working-group/src/lib.rs#L489-L493) there is no way to check if a given application is related to the opening that is being filled. **This also means that an opening can be filled by someone with less or no stakes than the required for the opening**

## Possible solutions
* Store in `Opening` the current applicants. This would make it easier(`O(number_of_applicants_for_opening)`) to check that a member isn't applying twice to the same opening as expected [here](https://github.com/Joystream/joystream/issues/2196#issuecomment-784189158). However, it would also be harder (`O(number_of_applicants_for_opening)`) to check if a set of applications are valid for a given opening.

* Store in the `Application` the corresponding opening. This would make it easier(`O(size_of_set)`) to check if a set of applications are valid for a given opening. But harder(`O(number_of_applications_in_storage)`) to check that a member isn't applying to the same opening twice.

### PR DESCRIPTION
**Pull request by conectado**
# Fix #2196 for `working_group` and  fix #2216 

In #2196 we decided to make the application for an opening require stake as an anti-spam mechanism for applications.

This PR includes:
* Make stake required for application in `working_group`
  * Fix tests and benchmarks
* Small fix for fee analysis to correctly calculate the minimum stake
* Store opening within application
  * When filling an opening verify the applications are consistent with the opening
  * Add tests for incorrect mapping between opening and application

**Note:** the corresponding handbook PR is Joystream/handbook#35

This also fixes Joystream/audits#14

### CODE REVIEW

File: runtime-modules/working-group/src/lib.rs [hunk 1]
```diff
@@ -492,28 +496,42 @@ decl_module! {
...
+                    !<CurrentLead<T,I>>::exists(),
+                    Error::<T, I>::CannotHireLeaderWhenLeaderExists
+                );
             }
 
-            let checked_applications_info = checks::ensure_succesful_applications_exist::<T, I>(&successful_application_ids)?;
+            let checked_applications_info =
+                checks::ensure_succesful_applications_exist::<T, I>(&successful_application_ids)?;
+
+            // Check that all applications are for the intended opening
```
Comment (shamil-gadelshin) [inline runtime-modules/working-group/src/lib.rs:525]: Consider using `ensure!` macro here.

File: runtime-modules/working-group/src/lib.rs [hunk 2]
```diff
@@ -414,48 +416,48 @@ decl_module! {
...
+                      &p.stake_parameters.staking_account_id
+                  ),
+                  Error::<T, I>::InvalidStakingAccountForMember
+            );
 
-                ensure!(
-                    T::StakingHandler::is_account_free_of_conflicting_stakes(&sp.staking_account_id),
-                    Error::<T, I>::ConflictStakesOnAccount
-                );
+              ensure!(
```
Comment (shamil-gadelshin) [inline runtime-modules/working-group/src/lib.rs]: Incorrect code indentation here.

Comment (shamil-gadelshin) [reply runtime-modules/working-group/src/lib.rs]: Also, line length greater than 99 (official rust recommendation)

File: runtime/src/lib.rs [hunk 3]
```diff
@@ -697,6 +697,10 @@ parameter_types! {
     pub const StorageWorkingGroupRewardPeriod: u32 = 14400 + 20;
     pub const ContentWorkingGroupRewardPeriod: u32 = 14400 + 30;
     pub const MembershipRewardPeriod: u32 = 14400 + 40;
+    // This should be more costly than `add_opening` fee with the current configuration
+    // the base cost of `add_opening` in tokens is 193. And has a very slight slope
+    // with the lenght with the length of rationale, with 2000 stake we are probably safe.
+    pub const MinimumStakeForOpening: u32 = 2000;
```
Comment (shamil-gadelshin) [inline runtime/src/lib.rs]: Incorrect type. Should be `Balance`.

File: runtime-modules/working-group/src/benchmarking.rs [hunk 4]
```diff
@@ -951,46 +899,16 @@ benchmarks_instance! {
                 1,
                 description,
                 OpeningType::Regular,
-                Some(stake_policy),
+                stake_policy,
                 Some(BalanceOf::<T>::max_value())
             ).into()
         );
     }
```
Comment (shamil-gadelshin) [inline runtime-modules/working-group/src/benchmarking.rs:907]: Why did you remove this branch? Also, in "leave_role" benchmark you could use "_" syntax extrinsic call.

**Review by shamil-gadelshin — CHANGES_REQUESTED** [review]
Looks good. Left several cosmetics comments.

File: runtime-modules/working-group/src/benchmarking.rs (hunk 4 above)
Comment (conectado) [reply runtime-modules/working-group/src/benchmarking.rs:907]: Because since the stake is now required there is no way to leave the role immediatly, which also means I should remove that branch from the extrinsic(Since no working group should have `MinimumStakeForOpening == 0` since that would prevent the stake from being an anti-spam mechanism and we require the leaving period for any application to be exclusively greater than `MinUnstakingPeriod` which means that no application can have an unstaking period of 0)

**Comment by bedeho** [comment]
Removing myself from review here.

**Review by shamil-gadelshin — APPROVED** [review]
LGTM
